### ISSUE TITLE
RDataReporting::residuals with adjoints

### ISSUE THREAD

**Issue by dweindl** [AMICI-dev/AMICI#2625]
So far, when using adjoint sensitivities, it's only possible to get `ReturnData::y` with `RDataReporting::full`. `RDataReporting::residuals` is not supported:

https://github.com/AMICI-dev/AMICI/blob/2d58f54be971b478a2318fea1324c74f61ff6610/src/solver.cpp#L738-L741

We should either allow that, and just skip the respective sensitivities (or basically just use `ReturnData::initializeResidualReporting(false)` in case of ASA), or introduce a new mode.

@FFroehlich : Any preference? 

CC @Doresic

**Comment by FFroehlich** [comment]
I agree that `RDataReporting::full` should give you everything that is available. `ReturnData::y` and `RDataReporting::residuals`  would be available but corresponding sensitivities are of course not. 

I don't quite get the relationship to the linked piece of code, it checks for `rdata_mode_ == RDataReporting::residuals`, so this does not apply for `RDataReporting::full`.

**Comment by dweindl** [comment]
> I agree that `RDataReporting::full` should give you everything that is available.

That's what's happening. All fine there

> I don't quite get the relationship to the linked piece of code, it checks for `rdata_mode_ == RDataReporting::residuals`, so this does not apply for `RDataReporting::full`.

That is also correct.

But I want to avoid using `RDataReporting::full`, because I only need `y` and `sllh`. The closest thing to that would be `RDataReporting::residuals`. But the referenced code forbids that.

**Comment by Doresic** [comment]
A nicer thing for my particular application (due to memory issues) would be to have another `amici.RDataReporting.observables` (naming up to change) that would give only llh, grad, y, and sigmay. 

However, this differs from `amici.RDataReporting.residuals` only by `res`, if I don't miss something. So that shouldn't be much more memory and should be a sufficient reduction from `amici.RDataReporting.full`.

**Comment by FFroehlich** [comment]
Yeah I agree that it sounds like introducing a new mode would be the way to go.

**Comment by dweindl** [comment]
New mode is fine for me. However, I am not sure if this makes too much sense:

https://github.com/AMICI-dev/AMICI/blob/2d58f54be971b478a2318fea1324c74f61ff6610/src/solver.cpp#L738-L741

Because the respective sensitivity members are only initialized for FSA:

https://github.com/AMICI-dev/AMICI/blob/2d58f54be971b478a2318fea1324c74f61ff6610/src/rdata.cpp#L81-L96

Removing that check above would be sufficient in my opinion, and we wouldn't need a new mode. But maybe I am missing something.

**Comment by FFroehlich** [comment]
> New mode is fine for me. However, I am not sure if this makes too much sense:
> 
> https://github.com/AMICI-dev/AMICI/blob/2d58f54be971b478a2318fea1324c74f61ff6610/src/solver.cpp#L738-L741
> 
> Because the respective sensitivity members are only initialized for FSA anyways and would just be skipped later on:
> 
> https://github.com/AMICI-dev/AMICI/blob/2d58f54be971b478a2318fea1324c74f61ff6610/src/rdata.cpp#L81-L96
> 
> Removing that check above would be sufficient in my opinion, and we wouldn't need a new mode. But maybe I am missing something.

Ah I see, yeah sounds plausible to simply remove that check.

**Comment by dweindl** [comment]
Unfortunately, it won't work that way. `sllh` is not computed then. I somehow thought, `RDataReporting::residuals` was a superset of `RDataReporting::likelihood`, but it's not.

Options
1) always including `sllh` with `RDataReporting::residuals` (maybe not so desirable, although probably negligible impact)
2) including `sllh` with `RDataReporting::residuals` if ASA is used (would make sense, otherwise there is no reason to use this combination)
3) new option `RDataReporting::observables_sllh`

I think, I'd go for option 2. No strong opinion, though.

**Comment by FFroehlich** [comment]
> Unfortunately, it won't work that way. `sllh` is not computed then. I somehow thought, `RDataReporting::residuals` was a superset of `RDataReporting::likelihood`, but it's not.
> 
> Options
> 
> 1. always including `sllh` with `RDataReporting::residuals` (maybe not so desirable, although probably negligible impact)
> 2. including `sllh` with `RDataReporting::residuals` if ASA is used (would make sense, otherwise there is no reason to use this combination)
> 3. new option `RDataReporting::observables_sllh`
> 4. supporting more or less arbitrary combinations via some flags (`reporting=Y|SIGMA|SLLH`, `reporting=SX`, ...)
> 
> I think, I'd go for option 2. No strong opinion, though.

the quick and dirty option would be to simply set `sllh` as `schi2`. Otherwise I would prefer option 3 as we would really blur the meaning of `RDataReporting` flags.
